### PR TITLE
 Fix DTensor Partial placement lost during autograd layout invariant (issue #180486)

### PR DESCRIPTION
**Pull request by anshul-si**
**Summary** Addresses https://github.com/pytorch/pytorch/issues/180486. When a Replicate DTensor parameter produces a non-contiguous Partial gradient (e.g. via transpose), autograd's clone_obey_contract calls new_empty_strided + copy_ to fix strides. new_empty_strided defaulted to Replicate placement, causing copy_ to trigger an unwanted all-reduce from Partial to Replicate.
                                                                                                                                              
Allow new_empty/new_empty_strided to inherit Partial placement from the input when shapes match, since uninitialized memory is immediately overwritten. Other factories (new_zeros, new_ones, new_full) keep the existing Replicate default to avoid incorrect values after reduction.     

**Test Cases**
1. pytest /data/users/anshulsi/pytorch/test/distributed/tensor/test_tensor_ops.py -k test_backward_partial_grad_with_transpose

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #180511

### CODE REVIEW

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 1]
```diff
@@ -282,7 +282,17 @@ def new_factory_strategy(op_schema: OpSchema) -> StrategyType:
...
+        # Uninitialized factories (new_empty*) can safely inherit Partial
+        # placement since their contents will be overwritten immediately
+        # (e.g., by autograd's clone_obey_contract: new_empty_strided + copy_).
+        is_uninitialized_factory = op_schema.op in (
+            aten.new_empty.default,
+            aten.new_empty_strided.default,
+        )
+        can_propagate_placement = input_spec.is_sharded() or (
+            is_uninitialized_factory and not input_spec.is_replicated()
+        )
```
Comment (pianpwk) [inline torch/distributed/tensor/_ops/_tensor_ops.py]: this logic reads confusing - is the correct logic: any sharding can propagate, and Partial can propagate for is_uninitialized_factory

**Review by pianpwk — APPROVED** [review]

File: test/distributed/tensor/test_tensor_ops.py [hunk 2]
```diff
@@ -1747,6 +1747,58 @@ def test_backward_no_allgather(self):
...
+        x = torch.randn(2, 8, device=self.device_type, requires_grad=True)
+        out = model(x)
+        out.sum().backward()
+
+        self.assertIsNotNone(model._grad_placement)
+        self.assertTrue(
+            all(isinstance(p, Partial) for p in model._grad_placement),
+            f"Expected Partial grad placement, got {model._grad_placement}",
+        )
+
```
Comment (pianpwk) [inline test/distributed/tensor/test_tensor_ops.py:1830]: nit: add a test that just calls new_empty/_strided?

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 3]
```diff
@@ -282,7 +282,17 @@ def new_factory_strategy(op_schema: OpSchema) -> StrategyType:
...
-        if tuple(input_shape) == tuple(output_shape) and input_spec.is_sharded():
+        # Uninitialized factories (new_empty*) can safely inherit Partial
+        # placement since their contents will be overwritten immediately
+        # (e.g., by autograd's clone_obey_contract: new_empty_strided + copy_).
+        is_uninitialized_factory = op_schema.op in (
+            aten.new_empty.default,
+            aten.new_empty_strided.default,
+        )
+        can_propagate_placement = input_spec.is_sharded() or (
+            is_uninitialized_factory and not input_spec.is_replicated()
```
Comment (zpcore) [inline torch/distributed/tensor/_ops/_tensor_ops.py]: Do we need the `not input_spec.is_replicated()` condition? I think this doesn't matter here.
